README
Verilog implementation of the HP saturn processor

licence: GPLv3 or later


timings:
           ___________   
reset:                |____________________________________________________
                ____      ____      ____      ____      ____      ____
clk :      ____|    |____|    |____|    |____|    |____|    |____|    |____
                          _________ _________ _________ _________ _________
counter:   ______________/____0____X____1____X____2____X____3____X____0____
                          _________                               _________
phase_0:   ______________|         |_____________________________|
                                    _________
phase_1:   ________________________|         |_____________________________
                                              _________
phase_2:   __________________________________|         |___________________
                                                        _________
phase_3:   ____________________________________________|         |_________

notes for using the ULX3S

Maybe linux ujprog won't find port because of insufficient priviledge. Either run ujprog as root or have udev rule:# this is for usb-serial tty device
SUBSYSTEM=="tty", ATTRS{idVendor}=="0403", ATTRS{idProduct}=="6015", \
MODE="664", GROUP="dialout"
this is for ujprog libusb access

ATTRS{idVendor}=="0403", ATTRS{idProduct}=="6015", \
GROUP="dialout", MODE="666"

history.txt
first delay is <async> -> posedge $glbnet$clk
second delay is posedge $glbnet$clk -> <async>

                    cells       speed 1     delay (a) 1 delay 1     arcs    speed 2     delay (a) 2 delay 2
2019-02-12 07:37      29        490.92MHz                2.68ns             330.14MHz               1.99ns
2019-02-12 07:48      29        490.92MHz   14.34ns      2.20ns             350.39MHz    4.14ns     1.77ns
2019-02-12 08:22      29        490.92MHz   14.34ns      2.68ns             330.58MHz    4.71ns     1.84ns
2019-02-12 08:39      29        490.92MHz   14.10ns      3.16ns        96   335.35MHz    4.46ns     1.93ns
2019-02-12 08:48      29        490.92MHz   14.10ns      3.16ns        96   335.35MHz    4.46ns     1.93ns
2019-02-12 11:11      29        490.92MHz   14.58ns      2.44ns        96   330.03MHz    4.74ns     2.04ns
2019-02-12 12:15      29        468.38MHz   14.34ns      2.68ns        96   330.58MHz    4.74ns     1.97ns
2019-02-12 12:48      48        242.95MHz   13.38ns      2.92ns       240   162.44MHz    4.13ns     2.05ns
2019-02-12 13:25      48        165.54MHz   14.58ns      3.64ns       240   176.46MHz    4.75ns     2.06ns
2019-02-12 13:28      48        142.84MHz   14.34ns      3.40ns       240   168.21MHz    4.71ns     2.31ns
2019-02-12 13:38      55        196.93MHz   14.58ns      3.16ns       343   166.53MHz    4.43ns     2.06ns
2019-02-12 14:51      50         84.74MHz   13.38ns      3.88ns       271   170.13MHz    4.15ns     2.35ns
2019-02-12 15:10      50         84.74MHz   13.38ns      3.88ns       271   170.13MHz    4.15ns     2.35ns
2019-02-12 15:39      50         75.52MHz   12.90ns      5.09ns       271   166.53MHz    4.14ns     2.55ns
2019-02-12 16:04      48        112.10MHz   14.34ns      2.92ns       241   163.99MHz    4.72ns     2.00ns
2019-02-12 17:27      56        142.90MHz   14.34ns      3.40ns       345   160.03MHz    4.95ns     2.25ns
2019-02-12 18:00      51        213.36MHz   13.98ns      2.92ns       273   165.43MHz    4.78ns     2.05ns
2019-02-12 22:07      49        112.10MHz   14.58ns      3.16ns       241   168.95MHz    4.74ns     1.93ns
2019-02-12 22:12      51        138.10MHz   15.06ns      2.92ns       273   154.61MHz    4.78ns     2.00ns
2019-02-12 22:20      49        112.10MHz   14.34ns      3.64ns       241   171.50MHz    4.55ns     2.33ns
2019-02-12 23:27      55        202.72MHz   13.14ns      4.61ns       344   162.63MHz    4.88ns     2.21ns
2019-02-13 07:46      48         98.83MHz   13.86ns      2.92ns       241   169.41MHz    4.11ns     1.92ns
2019-02-13 14:53      51        153.35MHz   16.74ns      3.40ns       273   155.47MHz    5.53ns     2.09ns
2019-02-13 23:21     318        113.77MHz   25.44ns     13.97ns      3061   117.75MHz    9.56ns     4.08ns
2019-02-14 09:00     353        118.23MHz   26.00ns     12.28ns      3334   119.40MHz   10.08ns     3.84ns
2019-02-14 22:11     403        115.09MHz   27.38ns     11.32ns      2430   111.51MHz   12.62ns     3.57ns
2019-02-15 07:14     535         97.16MHz   26.57ns      9.16ns      3073    96.10MHz   13.38ns     3.13ns
2019-02-15 11:07     577        102.13MHz   28.63ns     11.09ns      3270   101.77MHz   13.65ns     3.67ns
2019-02-15 17:12    1544         70.76MHz   33.39ns     12.77ns     10374    75.73MHz   16.74ns     3.99ns
2019-02-16 11:18    1552         72.00MHz   34.10ns     10.85ns     10072    75.65MHz   17.35ns     3.46ns     
2019-02-16 12:18    1453         62.31MHz   33.86ns     12.05ns     10323    70.66MHz   16.96ns     3.59ns
2019-02-16 22:54    1629         60.66MHz   33.62ns     12.05ns     11562    56.91MHz   17.11ns     3.84ns
2019-02-17 12:58    1679         68.43MHz   38.91ns     12.77ns     10906    70.19MHz   18.76ns     4.00ns
2019-02-17 15:11    1677         70.29MHz   34.92ns     13.01ns     11788    74.69MHz   17.42ns     3.88ns
2019-02-17 19:30    1637         74.68MHz   34.80ns     12.77ns     11687    68.49Mhz   18.03ns     4.01ns
2019-02-17 20:25    1733         72.37MHz   32.87ns     12.77ns     12213    73.52MHz   16.22ns     3.87ns
2019-02-17 21:21    1734         71.16MHz   32.10ns     12.05ns     11359    73.56MHz   17.19ns     3.73ns
2019-02-17 22:31    1573         77.32MHz   32.91ns     12.77ns     10265    70.53MHz   17.35ns     4.13ns
2019-02-17 22:48    1067         69.94MHz   32.87ns     12.77ns      6427    74.33MHz   17.71ns     3.96ns 
2019-02-17 23:04    1158         64.11MHz   37.87ns     12.77ns      7149    66.74MHz   19.03ns     4.06ns
2019-02-18 07:45    1128         74.65MHz   36.49ns     13.73ns      7586    75.72MHz   17.57ns     3.99ns
2019-02-19 16:26    1500         64.68MHz   42.16ns     17.32ns     10387    63.05MHz   19.43ns     7.69ns

mask_gen.v
module mask_gen (
// ports
    clk,
    nibble_width,
    nibble_start,
    mask
);
input clk; // clock
input wire [3:0] nibble_width; // length of mask in nibbles
input wire [3:0] nibble_start; // nibble where the mask starts
output reg [63:0] mask;// 64 bits mask

reg [4:0] n_max;
wire [3:0] nm1 = n_max[3:0];
reg [15:0] bitmask_1;
reg [15:0] bitmask_2;
reg [15:0] bitmask;
//wire [3:0] nm1;

always @( posedge clk) begin
    bitmask_1[ 0] = nibble_start==0;
    bitmask_1[ 1] = nibble_start==1  | bitmask_1[0];
    bitmask_1[ 2] = nibble_start==2  | bitmask_1[1];
    bitmask_1[ 3] = nibble_start==3  | bitmask_1[2];
    bitmask_1[ 4] = nibble_start==4  | bitmask_1[3];
    bitmask_1[ 5] = nibble_start==5  | bitmask_1[4];
    bitmask_1[ 6] = nibble_start==6  | bitmask_1[5];
    bitmask_1[ 7] = nibble_start==7  | bitmask_1[6];
    bitmask_1[ 8] = nibble_start==8  | bitmask_1[7]; 
    bitmask_1[ 9] = nibble_start==9  | bitmask_1[8];
    bitmask_1[10] = nibble_start==10 | bitmask_1[9];
    bitmask_1[11] = nibble_start==11 | bitmask_1[10];
    bitmask_1[12] = nibble_start==12 | bitmask_1[11];
    bitmask_1[13] = nibble_start==13 | bitmask_1[12];
    bitmask_1[14] = nibble_start==14 | bitmask_1[13];
    bitmask_1[15] = nibble_start==15 | bitmask_1[14];
    $display("bm1 : %b", bitmask_1);


    n_max <= nibble_start + nibble_width + 1;
    $display("n_max : %h", n_max);
    //nm1[3:0] = n_max[3:0];

    bitmask_2[15] = nm1==15;
    bitmask_2[14] = nm1==14 | bitmask_2[15];
    bitmask_2[13] = nm1==13 | bitmask_2[14];
    bitmask_2[12] = nm1==12 | bitmask_2[13];
    bitmask_2[11] = nm1==11 | bitmask_2[12];
    bitmask_2[10] = nm1==10 | bitmask_2[11];
    bitmask_2[ 9] = nm1==9  | bitmask_2[10];
    bitmask_2[ 8] = nm1==8  | bitmask_2[9];
    bitmask_2[ 7] = nm1==7  | bitmask_2[8];
    bitmask_2[ 6] = nm1==6  | bitmask_2[7];
    bitmask_2[ 5] = nm1==5  | bitmask_2[6];
    bitmask_2[ 4] = nm1==4  | bitmask_2[5];
    bitmask_2[ 3] = nm1==3  | bitmask_2[4];
    bitmask_2[ 2] = nm1==2  | bitmask_2[3];
    bitmask_2[ 1] = nm1==1  | bitmask_2[2];
    bitmask_2[ 0] = nm1==0  | bitmask_2[1];
    $display("bm2 : %b", bitmask_2);

    bitmask = n_max[4] ? bitmask_1 | bitmask_2 : bitmask_1 & bitmask_2;
    $display("bm  : %b", bitmask);

    mask[ 3: 0] = {4{bitmask[ 0]}}; 
    mask[ 7: 4] = {4{bitmask[ 1]}}; 
    mask[11: 8] = {4{bitmask[ 2]}}; 
    mask[15:12] = {4{bitmask[ 3]}}; 
    mask[19:16] = {4{bitmask[ 4]}}; 
    mask[23:20] = {4{bitmask[ 5]}}; 
    mask[27:24] = {4{bitmask[ 6]}}; 
    mask[31:28] = {4{bitmask[ 7]}}; 
    mask[35:32] = {4{bitmask[ 8]}}; 
    mask[39:36] = {4{bitmask[ 9]}}; 
    mask[43:40] = {4{bitmask[10]}}; 
    mask[47:44] = {4{bitmask[11]}}; 
    mask[51:48] = {4{bitmask[12]}}; 
    mask[55:52] = {4{bitmask[13]}}; 
    mask[59:56] = {4{bitmask[14]}}; 
    mask[63:60] = {4{bitmask[15]}}; 
end

endmodule

`ifdef SIM

//`timescale 1 ns / 100 ps

module mask_gen_tb;

// inputs
reg clock;

reg [3:0] nw;
reg [3:0] ns;
// outputs
wire [63:0] m;


mask_gen U0 (
    .clk (clock),
    .nibble_width (nw),
    .nibble_start (ns),
    .mask (m)
);


always 
    #10 clock = (clock === 1'b0);

initial begin
    //$monitor ("clk %b", clock);
    $monitor ("clk %b | nw %d | ns %d | m %h", clock, nw, ns, m);
    //#10 $display("1");
    //#10 $display("2");
    //#10 $finish;
end 


initial begin
    $dumpfile("text.vcd");
    $dumpvars(clock, nw, ns, m);
    $display($time, "starting simulation");
    clock = 0;
    $display("starting the simulation");
    run_mask_gen(4, 0);
    run_mask_gen(4, 1);
    run_mask_gen(4, 2);
    run_mask_gen(4, 3);
    run_mask_gen(4, 4);
    run_mask_gen(4, 5);
    run_mask_gen(4, 6);
    run_mask_gen(4, 7);
    run_mask_gen(4, 8);
    run_mask_gen(4, 9);
    run_mask_gen(4,10);
    run_mask_gen(4,11);
    run_mask_gen(4,12);
    run_mask_gen(4,13);
    run_mask_gen(4,14);
    run_mask_gen(4,15);

    //run_mask_gen(4, 0);
    //run_mask_gen(4, 0);
    //run_mask_gen(4, 0);
    //run_mask_gen(4, 0);
    $finish;
end


task run_mask_gen;
    input [3:0] _nw;
    input [3:0] _ns;
    begin
        $display("running", _nw, _ns);
        @(posedge clock);
        nw = _nw;
        ns = _ns;
    end
endtask

endmodule

`endif
